### ISSUE TITLE
Mounts are inside blue box (or grey or gold)

### ISSUE THREAD

**Issue by Tsspidermine** [tModLoader/tModLoader#1918]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Version

Latest public-1.4-alpha

### OS

Mac

### Platform

Steam

### Description

When I get on a mount such as the Drill Containment Unit, the texture for the mount is inside of a blue box
<img width="1440" alt="image" src="https://user-images.githubusercontent.com/51246489/138568622-ce4be546-ee70-4773-9b70-8cbd2cdf0d2d.png">

### Steps to reproduce

1. Get on mount by either using the item, or equipping it and pressing R

### Expected Behavior

There should be no blue box around the mount's texture

### Actual Behavior

There is a blue box enshrouding the mount

### Reproduction frequency

Every time I use my mount, I get the box.

**Comment by Jawell** [comment]
Duplicates:
#2158 
#2191

**Comment by Mirsario** [comment]
Issue seems to be common but exclusive to macs, aforementioned duplicates have extra info.
Seems like both of the provided logs use OpenGL rendering, need to check whether Vulkan/MoltenVK changes anything.
For anyone who wants to try, you need to start the game in a shell using:
`FNA3D_FORCE_DRIVER=Vulkan ./start-tModLoader.sh`

**Comment by Solxanich** [comment]
For those on Steam, you can also set the steam launch parameter to have:
/gldevice:Vulkan.

Steam Launch Parameter location shown below:
https://media.discordapp.net/attachments/871289059396448337/937463311405056000/unknown.png

**Comment by Jawell** [comment]
@Solxanich thank you, it's works for me

Can we implement something like auto choose rendering without launch parameters? Or this issue relevant only for tmodloader alpha version? Because, I don't see any problems in native terraria 1.4

**Comment by Solxanich** [comment]
An indirect fix was attempted, however not all Mac devices supported Vulkan.
The fix was reverted, and thus re-opened

**Comment by Goseale** [comment]
I have the same issue and /gldevice:Vulkan didnt work. I'm running PopOS 22.01 LTS with KDE Plasma

**Comment by Shadowblitz16** [comment]
I'm on linux (KDE neon 5.24) and my boxes are grey.
![image](https://user-images.githubusercontent.com/2453707/169920269-bbb00ff7-15a5-4230-8f4e-a30db3c5d680.png)

**Comment by Shadowblitz16** [comment]
Trying vulkan gives some strange visual bugs...
![image](https://user-images.githubusercontent.com/2453707/170184460-c3782c68-dbe9-4c65-be50-6877484fc2e3.png)

mounts work now tho haha.

**Comment by Rafii2198** [comment]
Sadly for me Vulkan didn't fix anything but also didn't break stuff unlike comment above. Everything stayed the same.
Using Linux and tModLoader v2022.5.103.9

**Comment by VioletRoses** [comment]
I tried forcing Vulkan and while it didn't introduce any new issues, it also didn't fix the issue (and changed the box color from green to white).
Running on Fedora 36 with tModLoader 2022.5.103.9, using Mesa-git drivers built on 5-28-2022

Edit: Disregard the comment about performance on Vulkan, was caused by something else entirely. After retesting, it seems that switching to Vulkan made the color box nearly transparent, but still visible.

**Comment by birb7** [comment]
Im on linux with the latest stable

for me, it does not go away when switching between vulkan and opengl. 
what it affects: 1/3rds of the starfarer sprites (always), all mounts (mod agnostic, regardless if mods are downloaded, disabled, deleted and etc.), and some boss animation frames (rarely)
the boxes are slightly more lime green when vulkan is used, and when using opengl the boxes are pure rgb green, white, or magenta. sometimes when switching to vulkan, 70% of the blocks go invisible, like dirt is gone but mud is not.

"gldevice:____"  - doesnt work
"gldevice:____ %command%" - doesnt work
"FNA3D_FORCE_DRIVER=OpenGL" - doesnt work
"FNA3D_FORCE_DRIVER=______ %command%" - does nothing except change what color the boxes are

Ive dealt with both
> Trying vulkan gives some strange visual bugs... ![image](https://user-images.githubusercontent.com/2453707/170184460-c3782c68-dbe9-4c65-be50-6877484fc2e3.png)
> 
> mounts work now tho haha.

and

> I tried forcing Vulkan and while it didn't introduce any new issues, it also didn't fix the issue (and changed the box color from green to white). Running on Fedora 36 with tModLoader 2022.5.103.9, using Mesa-git drivers built on 5-28-2022
> 
> Edit: Disregard the comment about performance on Vulkan, was caused by something else entirely. After retesting, it seems that switching to Vulkan made the color box nearly transparent, but still visible.

however, neither persisted and would change to green boxes upon restart

**Comment by Katknapp** [comment]
I also have this issue on Linux.

**Comment by tris790** [comment]
I had the same mount inside the RGB square bug on Linux Endeavour OS (Arch).

I fixed it using this launch flag in steam: FNA3D_FORCE_DRIVER=Vulkan %command%
The /gldevice:Vulkan flag didn't seem to work in my case but FNA3D_FORCE_DRIVER=Vulkan did the trick.
![image](https://user-images.githubusercontent.com/16129167/175866423-4840706f-9aee-49ce-9217-88cc43070551.png)

**Comment by kitsugo** [comment]
I have the same issue on Linux (Arch) and tried to do a bit of experimenting with this issue.

Using OpenGL, I noticed I actually get random results whenever I start the game. Sometimes mounts are entirely replaced by colored boxes, sometimes the boxes are slightly transparent, and sometimes all mounts display perfectly fine or only some with boxes only very slightly colored, so that it isn't really noticeable.

Unfortunately, however, I could not detect any striking difference within the log files in all these cases. Whether the mounts displayed correctly or not, the log file was exactly the same. I have tested switching to Vulkan, but this also did not resolve the issue for me.

The only interesting warning I sometimes received was a RangeException related to the `SurfaceBackgroundStylesLoader.DrawMiddleTexture()` method, however, it seems unrelated to this issue as even when it didn't appear, the texture glitch still persisted.

The full warning message:
```
[01:22:39] [.NET ThreadPool Worker/INFO] [tML]: Setting up...
[01:22:39] [Main Thread/WARN] [tML]: Silently Caught Exception: 
System.IndexOutOfRangeException: Index was outside the bounds of the array.
   at Terraria.ModLoader.SurfaceBackgroundStylesLoader.DrawMiddleTexture() in tModLoader\Terraria\ModLoader\BackgroundLoaders.cs:line 220
   at Terraria.Main.DrawSurfaceBG() in tModLoader\Terraria\Main.cs:line 46891
   at Terraria.Main.DrawBG() in tModLoader\Terraria\Main.cs:line 50540
   at Terraria.Main.DoDraw(GameTime gameTime) in tModLoader\Terraria\Main.cs:line 48904
   at Terraria.Main.Draw(GameTime gameTime) in tModLoader\Terraria\Main.cs:line 48494
   at Microsoft.Xna.Framework.Game.Tick() in D:\a\tModLoader\tModLoader\FNA\src\Game.cs:line 559
   at Microsoft.Xna.Framework.Game.RunLoop() in D:\a\tModLoader\tModLoader\FNA\src\Game.cs:line 886
   at Microsoft.Xna.Framework.Game.Run() in D:\a\tModLoader\tModLoader\FNA\src\Game.cs:line 415
   at Terraria.Program.LaunchGame_(Boolean isServer) in tModLoader\Terraria\Program.cs:line 229
   at Terraria.Program.LaunchGame(String[] args, Boolean monoArgs) in tModLoader\Terraria\Program.cs:line 194
   at MonoLaunch.<>c__DisplayClass2_0.<Main>b__0() in tModLoader\Terraria\MonoLaunch.cs:line 65
   at MonoLaunch.Main(String[] args) in tModLoader\Terraria\MonoLaunch.cs:line 72
```

**Comment by plasmacytoid** [comment]
Hay, I am having the same issue as you guys: I tried all the things talked about, but nothing helped, I hope someone finds a patch.

**Comment by Chicken-Bones** [comment]
Can someone send a world and player where it happens?

**Comment by Katknapp** [comment]
> Can someone send a world and player where it happens?
[World.zip](https://github.com/tModLoader/tModLoader/files/9088569/World.zip)

**Comment by lesliev** [comment]
I'm on Ubuntu Linux 20.04.4 and "FNA3D_FORCE_DRIVER=Vulkan %command%" didn't work for me.

**Comment by jellyfish-org** [comment]
I'm on mac, tried all the fixes and none of them worked :(

**Comment by Chicken-Bones** [comment]
Every person with this issue needs to post a `client.log` (and preferably a world _and_ player). Without seeing your logs it's impossible to tell what driver versions/graphics cards this occurs on. Many linux and mac users don't experience the issue and can't replicate it, so it's not platform universal.

If someone can figure out how to use renderdoc and do a frame dump, that would immensely help.

**Comment by kitsugo** [comment]
I have tested the issue again with both drivers (Vulkan, OpenGL) and compiled the log files as well as player and world.

Unfortunately, there is still no difference with using either driver for me. The colored boxes still have a random color on every new game launch and are on some occasions even slightly translucent. It should also be noted that dyes do affect the colored boxes and when a dye is used on a slightly translucent box, it can sometimes visually disappear that way.

I also noticed that for some reason the launch parameter `/gldevice:Vulkan` has no effect anymore and still starts the game with the OpenGL driver. This can be seen in the attached `transparent_client.log` file.

Using the attached world and player, the`opengl_client.log` file was produced by starting the game with `FNA3D_FORCE_DRIVER=OpenGL ./start-tModLoader.sh`and the `vulkan_client.log`file was produced by starting the game with `FNA3D_FORCE_DRIVER=Vulkan ./start-tModLoader.sh`.
In both instances the color box bug was present with no noticeable difference.

The `transparent_client.log` file was produced by starting the game with the `/gldevice:Vulkan` parameter in Steam, but (as mentioned) still used the OpenGL driver for some reason. However, this game launch produced a slightly transparent colorbox around the mounts which is different from the other two times.

[mount_graphicbug_logs.zip](https://github.com/tModLoader/tModLoader/files/9127682/mount_graphicbug_logs.zip)

**Comment by Rafii2198** [comment]
I never used RenderDoc, but I tried my best to do it, I captured 2 frames, one where it is just rendering me in a mount which was in that box, second when I came to mines to do same with minecart but with very similar errors that Shadowblitz16 got at the beginning of the issue. Also included client.log.
I used many mods which probably won't help as well as I don't know if I did the dump right way, apologies in advance. I will provide better info tomorrow with less mods as well as more files.

[render-errors.zip](https://github.com/tModLoader/tModLoader/files/9128731/render-errors.zip)
Had to use Vulkan with `FNA3D_FORCE_DRIVER=Vulkan` as got errors with RenderDoc when using OpenGL.

Also just speculation but that error when all the thing render incorrectly only happens on Vulkan and when I alt-tabbed, but I just tested it 2 times. As I said I'll make better testing and report tomorrow. 

Sorry for no response but I tried to find something interesting for several days but I couldn't find anything and I think that people below made better job anyways.

**Comment by MrLavender** [comment]
**EDIT - Spoke too soon, see next post**

Seems to be a .NET issue. Using a newer version fixes the problem for me.

```
$ cd ~/.local/share/Steam/steamapps/common/tModLoader
$ export LD_LIBRARY_PATH=~/.local/share/Steam/steamapps/common/tModLoader/Libraries/Native/Linux

$ dotnet/6.0.0/dotnet --info

Host (useful for support):
  Version: 6.0.0
  Commit:  4822e3c3aa

.NET SDKs installed:
  No SDKs were found.

.NET runtimes installed:
  Microsoft.NETCore.App 6.0.0 [/home/mrlavender/.local/share/Steam/steamapps/common/tModLoader/dotnet/6.0.0/shared/Microsoft.NETCore.App]

$ dotnet/6.0.0/dotnet tModLoader.dll
```

Has the mount issue. If I install latest .NET Core 6 SDK and use that instead the problem is gone.

```
$ cd ~/.local/share/Steam/steamapps/common/tModLoader
$ export LD_LIBRARY_PATH=~/.local/share/Steam/steamapps/common/tModLoader/Libraries/Native/Linux

$ dotnet --info
.NET SDK (reflecting any global.json):
 Version:   6.0.106
 Commit:    4f1e0330de

Runtime Environment:
 OS Name:     manjaro
 OS Version:
 OS Platform: Linux
 RID:         arch-x64
 Base Path:   /usr/share/dotnet/sdk/6.0.106/

Host (useful for support):
  Version: 6.0.6
  Commit:  7cca709db2

.NET SDKs installed:
  6.0.106 [/usr/share/dotnet/sdk]

.NET runtimes installed:
  Microsoft.NETCore.App 6.0.6 [/usr/share/dotnet/shared/Microsoft.NETCore.App]

$ dotnet tModLoader.dll
```

So as a temporary fix I edited ScriptCaller.sh to force using the system version;

```
#if [[ -f "$install_dir/dotnet" || -f "$install_dir/dotnet.exe" ]]; then
#  echo "Launched Using Local Dotnet"  2>&1 | tee -a "$LogFile"
#  [[ -f "$install_dir/dotnet" ]] && chmod a+x "$install_dir/dotnet"
#  exec "$install_dir/dotnet" tModLoader.dll "$@" 2>"$NativeLog"
#else
  echo "Launched Using System Dotnet"  2>&1 | tee -a "$LogFile"
  exec dotnet tModLoader.dll "$@" 2>"$NativeLog"
#fi
```
---

BTW!!! While investigating this issue I notice that .NET doesn't properly respect `LD_LIBRARY_PATH`. It will search that path but will prefer system libraries. I have `libFAudio.so` 2.0.0.7 on my system and it uses that instead of the version you ship. This doesn't cause any problems for me but seems like a massive gotcha for anyone with incompatible system versions of FAudio or FNA3D.

You can check what libraries the process is using with `pldd`;

```
$ ps ux | grep dotnet
mrlavender   12100 91.2  3.9 8605580 1309748 pts/1 Rl+  15:37   0:46 dotnet tModLoader.dll
mrlavender   12236  0.0  0.0   6688  2600 pts/2    S+   15:37   0:00 grep --colour=auto dotnet

$ sudo pldd 12100
12100:  /usr/share/dotnet/dotnet
/home/mrlavender/.local/share/Steam/steamapps/common/tModLoader/Libraries/Native/Linux/libSDL2.so
/home/mrlavender/.local/share/Steam/steamapps/common/tModLoader/Libraries/Native/Linux/libFNA3D.so.0
/home/mrlavender/.local/share/Steam/steamapps/common/tModLoader/Libraries/Native/Linux/libsteam_api64.so
/home/mrlavender/.local/share/Steam/linux64/steamclient.so
/usr/lib/libFAudio.so
:
:
and many other libraries
```

See https://github.com/dotnet/runtime/issues/34711

**Comment by MrLavender** [comment]
So I spoke too soon.

Running tModLoader like this with latest Net Core 6 SDK does fix the issue **most** of the time (80-90%).

```
$ cd ~/.local/share/Steam/steamapps/common/tModLoader
$ export LD_LIBRARY_PATH=~/.local/share/Steam/steamapps/common/tModLoader/Libraries/Native/Linux
$ dotnet tModLoader.dll
```

But starting it via `start-tModLoader.sh` makes the issue much more common again even if I force use of system dotnet (why does that script start two instances of tModLoader btw?)

```
$ ps ux | grep dotnet
mrlavender   15985 55.5  4.6 9759896 1538940 ?     Sl   10:19   2:38 dotnet tModLoader.dll
mrlavender   16062  0.0  0.2 3324664 94032 ?       Sl   10:19   0:00 /usr/share/dotnet/dotnet tModLoader.dll -terrariasteamclient 154
mrlavender   16255  0.0  0.0   6688  2596 pts/2    S+   10:23   0:00 grep --colour=auto dotnet
```

And even stranger... is that I've now noticed that the issue rarely doesn't appear even when using the default install...

So unfortunately the issue is not 100% reproducible one way or the other, it's just more or less common depending on how I launch the dll.

I can't see any difference in log files or pldd or anything else I can think of between when it's rendering correctly and when it's not. I can't get RenderDoc to work (it says "Context not created via CreateContextAttribs. Capturing disabled. Only OpenGL 3.2+ contexts are supported." - also does that with vanilla Terraria).

Am open to any other suggestions to try. Perhaps MESA/radeonsi debug flags might provide some info?

https://docs.mesa3d.org/envvars.html

Here is a zip containing client.log and the world+player I'm using to test (it's just a newly created small world and player, only thing I've done is given the player a Fuzzy Carrot mount. My testing procedure is start tModLoader, enter world, press R for mount, see if it renders correctly, save+exit game).

[mount-test.zip](https://github.com/tModLoader/tModLoader/files/9167306/mount-test.zip)

System info;

Operating System: Manjaro Linux
KDE Plasma Version: 5.24.6
KDE Frameworks Version: 5.96.0
Qt Version: 5.15.5
Kernel Version: 5.18.12-3-MANJARO (64-bit)
Graphics Platform: X11
Processors: 16 × AMD Ryzen 7 5700G with Radeon Graphics
Memory: 31.2 GiB of RAM
Graphics Processor: AMD Radeon RX 5500 XT

**Comment by MrLavender** [comment]
OK! Got a RenderDoc 1.20 capture in both working and failure states (needs FNA3D_OPENGL_FORCE_CORE_PROFILE=1 otherwise it defaults to compatibility profile 2.1 which was my problem before).

What's very interesting is that I couldn't trigger the issue when running under RenderDoc until I turned on the "Verify Buffer Access" option. Then it happens 100% of the time. So that seems like a good clue?

I guess my previous messing with different .NET version and launch method was just changing if/when this buffer problem occurred.

This zip contains;

tModLoader.cap - capture launch settings (you need to change the paths for your system)

failure.rcd - captured frame with the issue (verify buffer access = on)

success.rcd - captured frame without the issue (verify buffer access = off)

[renderdoc-capture.zip](https://github.com/tModLoader/tModLoader/files/9169815/renderdoc-capture.zip)

(World+player are in the zip I posted previously).

**Comment by Chicken-Bones** [comment]
Fixed in 765e98d8998a297a658b96952378a19baba8f085

**Comment by Chicken-Bones** [comment]
Thanks to @covers1624 @brandon3055 for debugging this with me
